shared/shared/postgresql_backend/pg_connection_handler.py

```diff
@@ -30,83 +33,105 @@ def __init__(self, connection_handler):
         self.connection_handler = connection_handler
 
     def __enter__(self):
-        self.connection_handler.connection.__enter__()
-        self.connection_handler.set_transaction_running(True)
+        self.connection = self.connection_handler.get_connection()
+        self.connection.__enter__()
 
     def __exit__(self, type, value, traceback):
-        self.connection_handler.set_transaction_running(False)
-        if self.connection_handler.connection:
-            self.connection_handler.connection.__exit__(type, value, traceback)
+        self.connection.__exit__(type, value, traceback)
+        self.connection_handler.put_connection(self.connection)
 
 
 @service_as_singleton
 class PgConnectionHandlerService:
 
-    connection: Optional[Any] = None
-    context = None
+    _storage: Optional[threading.local]
+    connection_pool: ThreadedConnectionPool
 
     environment: EnvironmentService
     shutdown_service: ShutdownService
 
     def __init__(self, shutdown_service: ShutdownService):
-        self.set_transaction_running(False)
         shutdown_service.register(self)
+        self._storage = None
+
+        min_conn = int(self.environment.try_get("DATASTORE_MIN_CONNECTIONS") or 1)
+        max_conn = int(self.environment.try_get("DATASTORE_MAX_CONNECTIONS") or 1)
+        self._semaphore = Semaphore(max_conn)
+        try:
+            self.connection_pool = ThreadedConnectionPool(
+                min_conn, max_conn, **self.get_connection_params()
+            )
+        except psycopg2.Error as e:
+            self.raise_error(
+                f"Database connection error ({type(e).__name__}) {e.pgcode}: {e.pgerror}"  # noqa
+            )
 
-    def set_transaction_running(self, value):
-        self.is_transaction_running = value
+    def get_current_connection(self):
+        if not self._storage:
+            self._storage = threading.local()
```

Comment:
are you setting a thread-scope object globally? `self._storage` is shared across all thready, if this singleton is created in the main thread. Maybe just drop the `self.` and just use local variables...

Reply:
Yes, but that's no problem. The `threading.local` has different attributes, depending ob which thread your are looking at it. In fact, after looking at it again, it sufficces to create it in the constructor and access it in the threads.